crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -706,9 +723,54 @@ where
                 is_star,
                 range: _,
             }) => {
+                // Save the state prior to visiting any of the `try` block.
+                //
+                // Potentially none of the `try` block could have been executed prior to executing
+                // the `except` block(s), so we will merge this state with all of the intermediate
+                // states during the `try` block before visiting the `except` blocks.
+                let pre_try_block_state = self.flow_snapshot();
+                let saved_definition_states = std::mem::take(&mut self.try_block_definition_states);
+                let visiting_nested_try_block = self.visiting_try_block;
+
+                // Visit the `try` block!
+                //
+                // If `visiting_try_block == true`, this informs `.add_definition()` to record the state
+                // in `self.try_block_definition_states` after each new definition is recorded.
+                self.visiting_try_block = true;
                 self.visit_body(body);
+                self.visiting_try_block = visiting_nested_try_block;
+
+                // Save the state immediately *after* visiting the `try` block
+                // but *before* we prepare for visiting the `except` block(s).
+                //
+                // We will revert to this state prior to visiting the the `else` block,
+                // as there necessarily must have been 0 `except` blocks executed
+                // if we hit the `else` block.
+                let post_try_block_state = self.flow_snapshot();
+
+                // Prepare for visiting the `except` block(s)
+                let visiting_try_block_states = std::mem::replace(
+                    &mut self.try_block_definition_states,
+                    saved_definition_states,
+                );
+                self.flow_restore(pre_try_block_state);
+                for state in visiting_try_block_states {
+                    self.flow_merge(state);
+                }
+                let pre_except_state = self.flow_snapshot();
+
+                // Account for the fact that we could be visiting a nested `try` block,
```

Comment:
I think the exception to this is if we can guarantee that the inner try block catches all exceptions. I think the only way we can be sure of this is if it has a bare except (or maybe an `except BaseException`?). In this case the outer try shouldn't see all the possible partial states from the inner try, it should only see the states after the except or else.

Reply:
Oh, and if the inner try has a finally block, the outer one should not see any state that doesn't include the finally block having run. So that's another thing we aren't quite doing right here yet. 

Good grief this is complex!

Reply:
> or maybe an `except BaseException`?

oh no... I don't know how we do that without type inference capabilities in the `SemanticIndexBuilder`?

Reply:
For now I'll just handle the bare `except:` case, I think :/

Reply:
> I don't know how we do that without type inference capabilities in the `SemanticIndexBuilder`?

Let's not worry about this wrinkle for this PR. I think handling bare except correctly is good; `except BaseException` is less common.

Eventually we will need to find a solution for adding conditional control flow paths that can be eliminated from consideration in type inference. It may be possible to do this as an extension of "constraints" using the Never type. We'll need this also for context manager `__exit__` return type handling.